src/core/solvers/box-qp.cpp

```diff
@@ -0,0 +1,258 @@
+  }
+
+  // Start the numerical iterations
+  for (std::size_t k = 0; k < maxiter_; ++k) {
+    clamped_idx_.clear();
+    free_idx_.clear();
+    // Compute the gradient
+    g_ = q;
+    g_.noalias() += H * x_;
+    for (std::size_t j = 0; j < nx_; ++j) {
+      const double& gj = g_(j);
+      const double& xj = x_(j);
+      const double& lbj = lb(j);
+      const double& ubj = ub(j);
+      if ((xj == lbj && gj > 0.) || (xj == ubj && gj < 0.)) {
+        clamped_idx_.push_back(j);
+      } else {
+        free_idx_.push_back(j);
+      }
+    }
+
+    // Check convergence
+    nf_ = free_idx_.size();
+    nc_ = clamped_idx_.size();
+    if (g_.lpNorm<Eigen::Infinity>() <= th_grad_ || nf_ == 0) {
+      if (k == 0) {  // compute the inverse of the free Hessian
+        Hff_.resize(nf_, nf_);
+        for (std::size_t i = 0; i < nf_; ++i) {
+          const std::size_t& fi = free_idx_[i];
+          for (std::size_t j = 0; j < nf_; ++j) {
+            Hff_(i, j) = H(fi, free_idx_[j]);
+          }
+        }
+        if (reg_ != 0.) {
+          Hff_.diagonal().array() += reg_;
+        }
+        Hff_inv_llt_ = Eigen::LLT<Eigen::MatrixXd>(nf_);
+        Hff_inv_llt_.compute(Hff_);
+        Eigen::ComputationInfo info = Hff_inv_llt_.info();
+        if (info != Eigen::Success) {
+          throw_pretty("backward_error");
+        }
+        Hff_inv_.setIdentity(nf_, nf_);
+        Hff_inv_llt_.solveInPlace(Hff_inv_);
+      }
+      solution_.Hff_inv = Hff_inv_;
+      solution_.x = x_;
+      solution_.free_idx = free_idx_;
+      solution_.clamped_idx = clamped_idx_;
+      return solution_;
+    }
+
+    // Compute the search direction as Newton step along the free space
+    qf_.resize(nf_);
+    xf_.resize(nf_);
+    xc_.resize(nc_);
+    dxf_.resize(nf_);
+    Hff_.resize(nf_, nf_);
+    Hfc_.resize(nf_, nc_);
+    for (std::size_t i = 0; i < nf_; ++i) {
+      const std::size_t& fi = free_idx_[i];
+      qf_(i) = q(fi);
+      xf_(i) = x_(fi);
+      for (std::size_t j = 0; j < nf_; ++j) {
+        Hff_(i, j) = H(fi, free_idx_[j]);
+      }
+      for (std::size_t j = 0; j < nc_; ++j) {
+        const std::size_t cj = clamped_idx_[j];
+        xc_(j) = x_(cj);
+        Hfc_(i, j) = H(fi, cj);
+      }
+    }
+    if (reg_ != 0.) {
+      Hff_.diagonal().array() += reg_;
+    }
+    Hff_inv_llt_ = Eigen::LLT<Eigen::MatrixXd>(nf_);
+    Hff_inv_llt_.compute(Hff_);
+    Eigen::ComputationInfo info = Hff_inv_llt_.info();
+    if (info != Eigen::Success) {
+      throw_pretty("backward_error");
+    }
+    Hff_inv_.setIdentity(nf_, nf_);
+    Hff_inv_llt_.solveInPlace(Hff_inv_);
+    dxf_ = -qf_;
+    if (nc_ != 0) {
+      dxf_.noalias() -= Hfc_ * xc_;
+    }
+    Hff_inv_llt_.solveInPlace(dxf_);
+    dxf_ -= xf_;
+    dx_.setZero();
+    for (std::size_t i = 0; i < nf_; ++i) {
+      dx_(free_idx_[i]) = dxf_(i);
+    }
+
+    // Try different step lengths
+    double fold = 0.5 * x_.dot(H * x_) + q.dot(x_);
```

Comment:
You may avoid some memory allocation here.

Reply:
done